docs/resources/azure_virtual_machine_datadisk.md

```diff
@@ -0,0 +1,197 @@
+   - `storage_account_type` if this is a managed disk what is the the storage type
+   - `id` the fully qualified id to the disk in Azure
+   - `subscription_id` if this is a managed disk the subscription that it is located in
+   - `resource_group` if this is a managed disk the resource group that it is in
+* `value` is the expected output fdrom the matcher
+
+The `count`, `has_data_disks?` and `has_managed_disks?` are catchall tests that give information about the virtual machine. The specific tests need to be used in conjunction with the `where` option as shown below.
+
+| Name        | Description                                                                                                         | Required | Example                           |
+|-------------|---------------------------------------------------------------------------------------------------------------------|----------|-----------------------------------|
+| group_name: | Azure Resource Group to be tested                                                                                   | yes      | MyResourceGroup                   |
+| name:       | Name of the Azure resource to test                                                                                  | no       | MyVM                              |
+| apiversion: | API Version to use when interrogating the resource. If not set then the latest version for the resoure type is used | no       | 2017-10-9                         |
+
+These options can also be set using the environment variables:
+
+ - `AZURE_RESOURCE_GROUP_NAME`
+ - `AZURE_RESOURCE_NAME`
+ - `AZURE_RESOURCE_API_VERSION`
+
+When the options have been set as well as the environment variables, the environment variables take priority.
+
+For example:
+
+```ruby
+describe azure_virtual_machine_data_disk(group_name: 'Inspec-Azure', name: 'Linux-Internal-VM') do
+  its('count') { should cmp > 0 }
+  it { should have_data_disks }
+end
+```
+
+## Matchers
+
+This InSpec audit resource has the following matchers:
+
+### eq
+
+Use the `eq` matcher to test the equality of two values: `its('Port') { should eq '22' }`.
+
+Using `its('Port') { should eq 22 }` will fail because `22` is not a string value! Use the `cmp` matcher for less restrictive value comparisons.
+
+**The following attributes are ones that are applied to the virtual machine itself and not specfic disks**
+
+### count
+
+Returns the number of data disks attached to the machine
+
+```ruby
+its('count') { should eq 1 }
+```
+
+### has_data_disks?
+
+Returns a boolean denoting if any data disks are attached to the machine
+
+```ruby
+it { should have_data_disks }
+```
+
+### has_managed_disks?
+
+Returns a boolean stating if the machine has Managed Disks for data disks.
+
+```ruby
+it { should have_managed_disks }
+```
+
+**The next set of attributes require the `where` operation to be used on the describe.**
+
+The following code shows an example of how to use the where clause.
```

Comment:
The only way to determine if a machine has data disks is to look at the VM, so yes this resource does find the VM and then provides tests for the selected data disk.
This _could_ be moved into the VM resource and a FilterTable setup for the datadisks, if that is less confusing (not that I think it is).
I took the decision to make it a separate resource so it was only data disks that were being tested. If I add it to the main VM resource this distinction is less obvious. Additionally it would mean that if Azure release a new feature to a VM that is array based then we would not be able to process it as would already be a FilterTable on the resource looking at the disks.
I do not think you can have more than one FilterTable, please advise if I am incorrect.

Reply:
I agree that data disks should be their own Inspec resource; I think your approach to split them off from the VM resource is an excellent one.

I still feel that the searching and properties seem a bit muddled to me, and I think it's because one might want to have both a singular and a plural resource for that (in other words, `azure_virtual_machine_datadisk` (intended to locate and test properties of a single datadisk) and `azure_virtual_machine_datadisks` (intended to test properties of all or a group of datadisks)).

That could be handled in a later PR, however.